Subject: [PATCH] git-instaweb: tolerate git worktree instances

If git instaweb is run from a git worktree instance then the
proper location to set up the gitweb directory is actually the
`git rev-parse --git-common-dir` directory (being careful to
work around bugs in that command by using --show-cdup first)
rather than the `git rev-parse --git-dir` directory.

To make git instaweb work reliably and predictably, modify
its behavior as follows:

1) first thing cd to the --show-cdup directory if any
2) get the full path to --git-common-dir
3) if --is-bare-repository continue as before but using the
   --git-common-dir instead of --git-dir
4) if .git does not exist or is not a directory or does not
   have a full path matching --git-common-dir's then cd to
   the --git-common-dir and treat the repository as bare

This combination works around the --git-common-dir bug while
making git instaweb work for pretty much any combination
of worktree / gitdir: desired.

Signed-off-by: Kyle J. McKay <mackyle@gmail.com>

---
 git-instaweb.sh | 16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

diff --git a/git-instaweb.sh b/git-instaweb.sh
index 88067731..8f4c0ce9 100755
--- a/git-instaweb.sh
+++ b/git-instaweb.sh
@@ -22,6 +22,20 @@ restart        restart the web server
 
 SUBDIRECTORY_OK=Yes
 . git-sh-setup
+cd "./$(git rev-parse --show-cdup)" || {
+	echo "Unable to determine path of top-level working tree directory" >&2
+	exit 1
+}
+gcd="$(git rev-parse --git-common-dir)" && [ -n "$gcd" ] && gcd="$(cd "$gcd" && pwd)" || {
+	echo "Unable to determine absolute path of git common directory" >&2
+	exit 1
+}
+GIT_DIR="$gcd"
+pseudo_bare=
+if ! [ -d .git ] || [ "$(cd .git && pwd -P)" != "$(cd "$gcd" && pwd -P)" ]; then
+	cd "$gcd"
+	pseudo_bare=1
+fi
 
 fqgitdir="$GIT_DIR"
 local="$(git config --bool --get instaweb.local)"
@@ -719,7 +733,7 @@ our \$projectroot = "$(dirname "$fqgitdir")";
 our \$git_temp = "$fqgitdir/gitweb/tmp";
 our \$projects_list = \$projectroot;
 EOF
-	if [ "$(git rev-parse --is-bare-repository)" = "true" ]; then
+	if [ -n "$pseudo_bare" ] || [ "$(git rev-parse --is-bare-repository)" = "true" ]; then
 		cat >> "$fqgitdir/gitweb/gitweb_config.perl" <<EOF
 our \$projects_list_restrict = "$(basename "$fqgitdir")";
 EOF
---
